### PR TITLE
BYOS - v0.1

### PR DESCRIPTION
**Pull request by alzix**
### Description
Add support for Bring Your Own Service (BYOS).
This PR will allow application developer to provide sources for `SPE_Target` within application directory.

Next steps: (not in this PR - but future releases)
[x] add code generation as a prebuilt step and remove committed autogenerated files
[x] generate sources under build directory rather than in tree
[x] switch order of the following folders for generated code TARGET_MBED_SPM/COMPONENT_SPE
[ ] NSPE -> SPE build dependency - trigger build automatically and remove committed HEX files

### Pull request type

    [ ] Fix
    [x] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [x] Breaking change

---
This change is [Reviewable](https://reviewable.io/reviews/kfnta/mbed-os/97)

### CODE REVIEW

**Comment by alzix** [comment]
CC @ndevillard

**Comment by alzix** [comment]
tested all kind of build:
regular tests
tests with PSA partitions
build app with secure partition sources https://github.com/kfnta/mbed-psa-actuator-poc/tree/alzix/byos
tomorrow CI tests and tests on a target

File: tools/build.py [hunk 1]
```diff
@@ -25,6 +25,7 @@
 
 
 # Be sure that the tools directory is in the search path
+
```
Comment (orenc17) [inline tools/build.py]: line

File: tools/build_api.py [hunk 2]
```diff
@@ -508,13 +508,22 @@ def prepare_toolchain(src_paths, build_dir, target, toolchain_name,
 )
 
 
+class SpeBuildType(object):
```
Comment (orenc17) [inline tools/build_api.py]: Move this to tools/psa/__init__.py

File: tools/project.py [hunk 3]
```diff
@@ -403,5 +411,6 @@ def main():
             exit(1)
     exit(0)
 
+
```
Comment (orenc17) [inline tools/project.py]: line

File: tools/psa/generate_partition_code.py [hunk 4]
```diff
@@ -27,149 +27,102 @@
 
 from tools.psa.mbed_spm_tfm_common import validate_partition_manifests, \
     manifests_discovery, parse_manifests, generate_source_files, \
-    MBED_OS_ROOT, SERVICES_DIR, TESTS_DIR
+    MBED_OS_ROOT
```
Comment (orenc17) [inline tools/psa/generate_partition_code.py]: can you check if it fits in one line now?

File: tools/resources/__init__.py [hunk 5]
```diff
@@ -594,7 +595,17 @@ def scan_with_config(self, src_paths, config):
         config.load_resources(self)
         return self
 
-    def filter_spe(self):
+    def filter_spe_only(self):
         spe_filter = lambda x: 'COMPONENT_SPE' in x
         for type in [FileType.ASM_SRC, FileType.C_SRC, FileType.CPP_SRC]:
             self._file_refs[type] = set([f for f in self._file_refs[type] if spe_filter(f.name) or spe_filter(f.path)])
+
+    def filter_spe_app(self):
```
Comment (orenc17) [inline tools/resources/__init__.py]: unify this and filter_spe_only into one function that receives a parameter(filter type)

**Comment by alzix** [comment]
___
*[tools/psa/release.py, line 196 at r1](https://reviewable.io/reviews/kfnta/mbed-os/97#-LckudglDN1yp6Z4NUiu:-LckudglDN1yp6Z4NUiv:bhzflwy) ([raw file](https://github.com/kfnta/mbed-os/blob/24825d2d05e06133f1d355e31f075707bb0a78aa/tools/psa/release.py#L196)):*
> ```Python
>         cmd += ['--app-config', TFM_MBED_APP]
>     else:
>         cmd += ['--artifact-name', 'psa_release_1.0']
> ```

why do we need --artifact-name only for SPM builds?
where 1.0 came from?

**Pull request closed without merging**